<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción--" data-toc-modified-id="Introducción---1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción  <a id="intro" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Objetivo:-" data-toc-modified-id="Objetivo:--1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Objetivo: <a id="objetivo" rel="nofollow"></a></a></span></li><li><span><a href="#Etapas-" data-toc-modified-id="Etapas--1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Etapas <a id="etapas" rel="nofollow"></a></a></span></li></ul></li><li><span><a href="#Etapa-1.-Descripción-de-los-datos-" data-toc-modified-id="Etapa-1.-Descripción-de-los-datos--2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Etapa 1. Descripción de los datos <a id="data_review" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Escribe-observaciones-de-tu-parte." data-toc-modified-id="Escribe-observaciones-de-tu-parte.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Escribe observaciones de tu parte.</a></span><ul class="toc-item"><li><span><a href="#Respuesta:" data-toc-modified-id="Respuesta:-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Respuesta:</a></span></li></ul></li></ul></li><li><span><a href="#Etapa-2.-Preprocesamiento-de-datos-" data-toc-modified-id="Etapa-2.-Preprocesamiento-de-datos--3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Etapa 2. Preprocesamiento de datos <a id="data_preprocessing" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Estilo-del-encabezado-" data-toc-modified-id="Estilo-del-encabezado--3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Estilo del encabezado <a id="header_style" rel="nofollow"></a></a></span></li><li><span><a href="#Valores-ausentes-" data-toc-modified-id="Valores-ausentes--3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Valores ausentes <a id="missing_values" rel="nofollow"></a></a></span></li><li><span><a href="#Duplicados-" data-toc-modified-id="Duplicados--3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Duplicados <a id="duplicates" rel="nofollow"></a></a></span></li><li><span><a href="#Tus-observaciones-" data-toc-modified-id="Tus-observaciones--3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Tus observaciones <a id="data_preprocessing_conclusions" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Observaciones" data-toc-modified-id="Observaciones-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Observaciones</a></span></li></ul></li></ul></li><li><span><a href="#Etapa-3.-Prueba-de-hipótesis-" data-toc-modified-id="Etapa-3.-Prueba-de-hipótesis--4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Etapa 3. Prueba de hipótesis <a id="hypothesis" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades-" data-toc-modified-id="Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades--4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id="activity" rel="nofollow"></a></a></span></li><li><span><a href="#Conclusiones" data-toc-modified-id="Conclusiones-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Conclusiones</a></span></li></ul></li><li><span><a href="#Conclusiones-" data-toc-modified-id="Conclusiones--5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusiones <a id="end" rel="nofollow"></a></a></span></li></ul></div>

# ¡Hola Carlos!

Mi nombre es Ezequiel Ferrario, soy code reviewer en Tripleten y tengo el agrado de revisar el proyecto que entregaste.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión.

Encontrarás mis comentarios más abajo - **por favor, no los muevas, no los modifiques ni los borres**.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta. Se aceptan uno o dos comentarios de este tipo en el borrador, pero si hay más, deberá hacer las correcciones. Es como una tarea de prueba al solicitar un trabajo: muchos pequeños errores pueden hacer que un candidato sea rechazado.
</div>

<div class="alert alert-block alert-danger">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>

Puedes responderme de esta forma:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>

Hola, muchas gracias por tus comentarios y la revisión.        
</div>

¡Empecemos!

-----------


<div class="alert alert-block alert-danger">
<b>Comentario general #1</b> <a class="tocSkip"></a>

Carlos, has realiza un excelente trabajo donde abarcaste todos los puntos del mismo.

A nivel codigo te supiste manejar con todos los metodos para extraer la informacion. En cuanto a tus analisis destacaran durante todo el proyecto ya que agregaron claridad en todo momento.

Resta un detalle por corregir.


Quedo atento a tu correccion, saludos.</div>


<div class="alert alert-block alert-success">
<b>Comentario general #2</b> <a class="tocSkip"></a>

Carlos, abarcado el unico detalle que faltaba y sumado a la excelente tabla de contenidos, el proyecto queda **aprobado**.

Exitos en lo que viene.

Saludos.</div>

----------

# Déjame escuchar la música

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción--" data-toc-modified-id="Introducción---1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción  <a id="intro" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Objetivo:-" data-toc-modified-id="Objetivo:--1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Objetivo: <a id="objetivo" rel="nofollow"></a></a></span></li><li><span><a href="#Etapas-" data-toc-modified-id="Etapas--1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Etapas <a id="etapas" rel="nofollow"></a></a></span></li></ul></li><li><span><a href="#Etapa-1.-Descripción-de-los-datos-" data-toc-modified-id="Etapa-1.-Descripción-de-los-datos--2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Etapa 1. Descripción de los datos <a id="data_review" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Escribe-observaciones-de-tu-parte." data-toc-modified-id="Escribe-observaciones-de-tu-parte.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Escribe observaciones de t
    u parte.</a></span><ul class="toc-item"><li><span><a href="#Respuesta:" data-toc-modified-id="Respuesta:-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Respuesta:</a></span></li></ul></li></ul></li><li><span><a href="#Etapa-2.-Preprocesamiento-de-datos-" data-toc-modified-id="Etapa-2.-Preprocesamiento-de-datos--3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Etapa 2. Preprocesamiento de datos <a id="data_preprocessing" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Estilo-del-encabezado-" data-toc-modified-id="Estilo-del-encabezado--3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Estilo del encabezado <a id="header_style" rel="nofollow"></a></a></span></li><li><span><a href="#Valores-ausentes-" data-toc-modified-id="Valores-ausentes--3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Valores ausentes <a id="missing_values" rel="nofollow"></a></a></span></li><li><span><a href="#Duplicados-" data-toc-modified-id="Duplicados--3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Duplicados <a id="duplicates" rel="nofollow"></a></a></span></li><li><span><a href="#Tus-observaciones-" data-toc-modified-id="Tus-observaciones--3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Tus observaciones <a id="data_preprocessing_conclusions" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Observaciones" data-toc-modified-id="Observaciones-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Observaciones</a></span></li></ul></li></ul></li><li><span><a href="#Etapa-3.-Prueba-de-hipótesis-" data-toc-modified-id="Etapa-3.-Prueba-de-hipótesis--4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Etapa 3. Prueba de hipótesis <a id="hypothesis" rel="nofollow"></a></a></span><ul class="toc-item"><li><span><a href="#Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades-" data-toc-modified-id="Hipótesis:-comparar-el-comportamiento-del-usuario-o-la-usuaria-en-las-dos-ciudades--4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id="activity" rel="nofollow"></a></a></span></li><li><span><a href="#Conclusiones" data-toc-modified-id="Conclusiones-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Conclusiones</a></span></li></ul></li><li><span><a href="#Conclusiones-" data-toc-modified-id="Conclusiones--5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusiones <a id="end" rel="nofollow"></a></a></span></li></ul></div>

<div class="alert alert-block alert-warning">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien la tabla de contenidos pero debe estar linkeada a las secciones (al clickear debe llevarnos a esa seccion) de esta menera es mas facil desplazarse.

Como consejo, si realizas bien todas las secciones (con su respectivo #) podes generarlo automáticamente desde jupyter lab. Para hacerlo, en la pestaña de herramientas de jupyter lab clickeas en el **botón de los puntos y barras**  (Table of contents) te generara automáticamente una tabla de contenidos linkeable y estética. A la **derecha** del seleccionador del tipo de celda (code/markdown)
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido, excelente.
</div>

## Introducción  <a id='intro'></a>

Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo: <a id='objetivo'></a>
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas <a id='etapas'></a>
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

# Etapas

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
import pandas as pd# Importar pandas


Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [2]:
# Leer el archivo y almacenarlo en df
df= pd.read_csv("/datasets/music_project_en.csv")


Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Esta bien. Pero en este caso remplaza el print() y utiliza el metodo sin nada (llamalo solo). Es decir llamas a la variable y al metodo correspondiente y ejecutas.

- *El ejemplo seria: nombre_df.metodo(10)*

Esto hara que sea mas legible y estetico.</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>

Corregido.
</div>

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;
- `'Track'`: título de la canción;
- `'artist'`: nombre del artista;
- `'genre'`: género de la pista;
- `'City'`: ciudad del usuario o la usuaria;
- `'time'`: la hora exacta en la que se reprodujo la canción;
- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. **No se usa el "_" para separa las palabras**.





### Escribe observaciones de tu parte. 
**Estas son algunas de las preguntas que pueden ser útiles:**



`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

#### Respuesta:

1. Todos los datos de las calumnas del DataFrame son "object" y los nombres de las columnas indicas el tipo de dato que almacenan 
2. Hay 65079 datos, por lo que si se pueden hacer hipoteisis basados en esta informacion, incluso despues de hacer  la limpieza dee datos.
2. Todos los datos de las celdas son correctos, se puede evidenciar filas duplicadas y ausencia de valores en las columnas `'Track'`,`'artist'`y`'genre'`

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelentes observaciones.</div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
columns = df.columns# Muestra los nombres de las columnas
print (columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
new_columns = []
for lower in columns:
    lowered = lower.lower()
    new_columns.append(lowered)
new_columns = new_columns    # Bucle en los encabezados poniendo todo en minúsculas
    
print(new_columns)

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
new_columns_2 = []
for space in new_columns:
    spaces = space.strip()
    new_columns_2.append(spaces) 
new_columns_2 = new_columns_2    
print (new_columns_2) # Bucle en los encabezados eliminando los espacios


['userid', 'track', 'artist', 'genre', 'city', 'time', 'day']


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
df.columns= new_columns_2

fin_columns = {"userid": "user_id"}
df.rename(columns=fin_columns, inplace = True)# Cambiar el nombre de la columna "userid"

print (df.columns)
print()
print(df.info())

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    63736 non-null  object
 2   artist   57512 non-null  object
 3   genre    63881 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
print (df.columns)# Comprobar el resultado: la lista de encabezados


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


[Volver a Contenidos](#back)

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien cambiando los encabezados.</div>

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
print(df.isna().sum())# Calcular el número de valores ausentes


user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
replace = ["track","artist"]
replace_1 =["genre"]
for new in replace:
    df[new].fillna("unknown",inplace = True)# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
for new_1 in replace_1:
    df[new_1].fillna("generic",inplace = True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


**Los valores ausentes de la columna "genero" son 1.198, que signica 1,84% del total de los datos, podrian tomarse como un margen de error, sin embargo tenerlos en cuenta para los calculos.**

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
print(df.info())# Contar valores ausentes


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  65079 non-null  object
 1   track    65079 non-null  object
 2   artist   65079 non-null  object
 3   genre    65079 non-null  object
 4   city     65079 non-null  object
 5   time     65079 non-null  object
 6   day      65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None


<div class="alert alert-block alert-success">

<b>Comentario del revisor</b> <a class="tocSkip"></a>

Buen bucle para remplazar los nulos.</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [13]:
print(df.duplicated().sum())# Contar duplicados explícitos


3826


Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
df = df.drop_duplicates()
# Eliminar duplicados explícitos


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
print(df.duplicated().sum())# Comprobar de nuevo si hay duplicados


0


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
print(df["genre"].sort_values().unique())# Inspeccionar los nombres de géneros únicos


['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien. Como consejo extra, contando la cantidad de elementos unicos - nunique() - que tenemos (antes y despues) es  un parametro para verificar si el error esta corregido o no.

Si tengo menos registros que antes estara corregido.</div>

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
wrong_genres= ["hip","hop","hip-hop"]# Función para reemplazar duplicados implícitos
correct_genre= "hiphop"
def replace_wrong_genres(dataframe, column,wrong_genres,correct_genre):
    for wrong in wrong_genres:
        dataframe[column].replace(wrong_genres,correct_genre,inplace=True)
    return df
        

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
replace_wrong_genres(df,"genre",wrong_genres,correct_genre)

# Eliminar duplicados implícitos


user_id                              track            artist  \
0      FFB692EC                  Kamigata To Boots  The Mass Missile   
1      55204538        Delayed Because of Accident  Andreas Rönnberg   
2        20EC38                  Funiculì funiculà       Mario Lanza   
3      A3DD03C9              Dragons in the Sunset        Fire + Ice   
4      E2DC1FAE                        Soul People        Space Echo   
...         ...                                ...               ...   
65074  729CBB09                            My Name            McLean   
65075  D08D4A55  Maybe One Day (feat. Black Spade)       Blu & Exile   
65076  C5E3A0D5                          Jalopiina           unknown   
65077  321D0506                      Freight Train     Chas McDevitt   
65078  3A64EF84          Tell Me Sweet Little Lies      Monica Lopez   

            genre         city      time        day  
0            rock  Shelbyville  20:28:33  Wednesday  
1            rock  Springfield  14:07:09     Friday  
2             pop  Shelbyville  20:58:07  Wednesday  
3            folk  Shelbyville  08:37:09     Monday  
4           dance  Springfield  08:34:34     Monday  
...           ...          ...       ...        ...  
65074         rnb  Springfield  13:32:28  Wednesday  
65075      hiphop  Shelbyville  10:00:00     Monday  
65076  industrial  Springfield  20:09:26     Friday  
65077        rock  Springfield  21:43:59     Friday  
65078     country  Springfield  21:59:46     Friday  

[61253 rows x 7 columns]

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos
print(df["genre"].sort_values().unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'ambient' 'americana' 'animated' 'anime' 'arabesk' 'arabic' 'arena'
 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé' 'baile' 'balkan'
 'beats' 'bigroom' 'black' 'bluegrass' 'blues' 'bollywood' 'bossa'
 'brazilian' 'breakbeat' 'breaks' 'broadway' 'cantautori' 'cantopop'
 'canzone' 'caribbean' 'caucasian' 'celtic' 'chamber' 'children' 'chill'
 'chinese' 'choral' 'christian' 'christmas' 'classical' 'classicmetal'
 'club' 'colombian' 'comedy' 'conjazz' 'contemporary' 'country' 'cuban'
 'dance' 'dancehall' 'dancepop' 'dark' 'death' 'deep' 'deutschrock'
 'deutschspr' 'dirty' 'disco' 'dnb' 'documentary' 'downbeat' 'downtempo'
 'drum' 'dub' 'dubstep' 'eastern' 'easy' 'electronic' 'electropop' 'emo'
 'entehno' 'epicmetal' 'estrada' 'ethnic' 'eurofolk' 'european'
 'experimental' 'extrememetal' 'fado' 'film' 'fitness' 'flamenco' 'folk'
 'folklore' 'folkmetal' 'folkrock' 'folktronica' 'forró' 'frankreich'
 'französisch' 

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien.</div>

**Organicé los indices de los datos**

In [20]:
df=df.reset_index(drop=True)
print(df.duplicated().sum())
df.info()

0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61253 entries, 0 to 61252
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  61253 non-null  object
 1   track    61253 non-null  object
 2   artist   61253 non-null  object
 3   genre    61253 non-null  object
 4   city     61253 non-null  object
 5   time     61253 non-null  object
 6   day      61253 non-null  object
dtypes: object(7)
memory usage: 3.3+ MB


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Perfecto restableciendo el indice.</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`

#### Observaciones 
Se organizaron los datos, hallamos el cantidad de elementos duplicados y se eliminaron 3.826 datos duplicados, ahora vamos atrabajar con 61.253 datos para comprobar las hipotesis y llegar a conclusiones. 

Adicionalmete se corriegueron algunos valores incorrecto dentro de las celdad que querian indicar el mismo genero. 
despues de la limpieza de los datos del DataFrame tenemos la cantidad optima para hacer las operaciones.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buenas observaciones.</div>

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [21]:
print(df["city"].sort_values().unique())
print()
dfc= df.groupby(by="city")['day'].count()# Contar las canciones reproducidas en cada ciudad

print(dfc)

['Shelbyville' 'Springfield']

city
Shelbyville    18512
Springfield    42741
Name: day, dtype: int64


**Observaciones:**
Despues del calculo se puede ver que en la ciudad de Springfield se escuchó 42.471 canciones, poco mas de el doble que Shelbyville con 18.512.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Perfecto.</div>

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [22]:

dfd = df.groupby(by="day")["city"].count()# Calcular las canciones reproducidas en cada uno de los tres días
print (dfd)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: city, dtype: int64


**Observaciones:**
Despues del calculo, los dias que más se escucha musica son los viernes con 21.840 caciones reproducidas mientras que los lunes se escuchan 486 cancines menos que el viernes (21.354). Por ultimo tenemos 18.059 reproducciones el miercoles dusminuyendo un 6,17% con respecto al lunes.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [23]:
def number_tracks(df, day, city ): # Declara la función number_tracks() con dos parámetros: day= y city=.
    info=df[df["day"]== day] # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=
    info_2=info[info["city"]== city] # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    info_3=info_2["user_id"].count() # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    return info_3 # Devolve el número de valores de la columna 'user_id'




<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buena la funcion.

Como recomendacion para futuros proyectos, podrias utilizar expresiones anidadas para que tu codigo sea mucho mas corto, por ejemplo si queremos que se cumplan dos condiciones consecutivas podemos utilizar el AND y en una linea de codigo resolvemos el mismo problema.</div>

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [24]:
print(number_tracks(df, "Monday", "Springfield"))# El número de canciones reproducidas en Springfield el lunes


15740


In [25]:
print(number_tracks(df,"Monday","Shelbyville"))# El número de canciones reproducidas en Shelbyville el lunes


5614


In [26]:
print(number_tracks(df,"Wednesday","Springfield"))# El número de canciones reproducidas en Springfield el miércoles


11056


In [27]:
print(number_tracks(df,"Wednesday","Shelbyville"))# El número de canciones reproducidas en Shelbyville el miércoles


7003


In [28]:
print(number_tracks(df,"Friday","Springfield"))# El número de canciones reproducidas en Springfield el viernes


15945


In [29]:
print(number_tracks(df,"Friday","Shelbyville"))# El número de canciones reproducidas en Shelbyville el viernes


5895


In [30]:
df_city = df[df["city"]=="Springfield"]
df_day=df_city.groupby(by="day")["user_id"].count()
print ("Springfield by",df_day.sort_values(ascending=False))
print ("Total",df_day.sum())
print()
df_city = df[df["city"]=="Shelbyville"]
df_day=df_city.groupby(by="day")["user_id"].count()
print ("Shelbyville by",df_day.sort_values(ascending=False))
print ("Total",df_day.sum())

Springfield by day
Friday       15945
Monday       15740
Wednesday    11056
Name: user_id, dtype: int64
Total 42741

Shelbyville by day
Wednesday    7003
Friday       5895
Monday       5614
Name: user_id, dtype: int64
Total 18512


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy bien creando estas agrupaciones para que quede aun mas claro.</div>

### Conclusiones

Hipótesis:
La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

La hipotesis es parcialmente correcta, la reproduccion de canciones difiere totalmente entre las dos ciudades, cualquera de los tres dias; sin embargo las dos ciudades tienen las misma tendencias los dias lunes y viernes en diferente proporciones, pero el miercoles difieren las dos ciudades es su tendencia.




[Volver a Contenidos](#back)

## Conclusiones <a id='end'></a>

1. En Springfield se escuchas más mucica los viernes, mientras que en Shelbyville es el miercoles.
2. La direfencia entre las reproducciones del lunes y el viernes es muy pequeña en las dos ciudades.
3. En Springfield se escucha más del doble de canciones que en Shelbyville.
4. En Springfield los lunes y viernes se escucha casi tres veces que en el Shelbyville.

De aqui podemos tomar varias decisiones segun la estrategia del area.

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelentes conclusiones, muy ordenadas y bien utilizado el bullet para cada item.</div>

**Nota**

En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)